server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,112 @@
+        this.client = client;
+    }
+
+    private record NodeTransportVersionTask(String nodeId, TransportVersion version) implements ClusterStateTaskListener {
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for node {} to cluster state", nodeId, e);
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                var t = c.getTask();
+                // this node's transport version might have been updated already/node has gone away
+                if (Objects.equals(builder.transportVersions().get(t.nodeId()), INFERRED_VERSION)) {
+                    builder.putTransportVersion(t.nodeId(), t.version());
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster() == false) return; // only if we're master
+
+        // if the min node version >= 8.8.0, and the min transport version is inferred,
+        // then refresh all inferred transport versions to their real versions
+        // now that everything should understand cluster state with transport versions
+        if (event.state().nodes().getMinNodeVersion().onOrAfter(Version.V_8_8_0)
+            && event.state().getMinTransportVersion().equals(INFERRED_VERSION)) {
+            // find all the relevant nodes
+            String[] nodes = event.state()
+                .transportVersions()
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue().equals(INFERRED_VERSION))
+                .map(Map.Entry::getKey)
+                .toArray(String[]::new);
+
+            NodesInfoRequest request = new NodesInfoRequest(nodes);
+            request.clear().addMetric(ClusterState.Metric.NODES.toString());
+            client.nodesInfo(request, new ActionListener<>() {
+                @Override
+                public void onResponse(NodesInfoResponse response) {
+                    for (NodeInfo n : response.getNodes()) {
+                        String nodeId = n.getNode().getId();
+                        masterService.submitTask(
+                            "update-transport-version[" + nodeId + "]",
+                            new NodeTransportVersionTask(nodeId, n.getTransportVersion()),
+                            null
+                        );
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    Log.warn("Could not read transport versions from nodes", e);
```

Comment:
Do we want to retry here? An error reading node info indicates something is up with the cluster. This listener will keep triggering on cluster state changes until it succeeds, so there'll be implicit retries there...

Reply:
We should protect against firing several of these requests concurrently (e.g. if a flurry of cluster state updates comes in just after the upgrade). And yes conversely IMO we should retry on failure because there may be no further cluster state updates for some unknown (long) amount of time too.

Reply:
Note that calling this `onFailure` indicates a top-level failure which is unlikely (maybe only happens if the node is shutting down). There's also `NodesInfoResponse#failures()`.

Reply:
For concurrent requests, I don't think we want to just drop newer ones if one's in progress, as there could be node updates in there we need to worry about. So they'll need to be queued up somewhere & fired off one by one...

Reply:
sounds like a job for `SingleResultDeduplicator`

Reply:
Whilst not directly usable, I've taken its ideas and added something similar directly to the listener. This should reduce the number of node info calls it makes when there's lots of state updates all at once